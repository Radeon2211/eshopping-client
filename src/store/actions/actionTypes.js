export const LOGIN_USER = 'LOGIN_USER';
export const LOGOUT_USER = 'LOGOUT_USER';
export const SET_PLACED_ORDERS = 'SET_PLACED_ORDERS';
export const SET_SELL_HISTORY = 'SET_SELL_HISTORY';
export const SET_PROFILE = 'SET_PROFILE';
export const UPDATE_TRANSACTION = 'UPDATE_TRANSACTION';
export const SET_PRODUCTS = 'SET_PRODUCTS';
export const SET_PRODUCT_DETAILS = 'SET_PRODUCT_DETAILS';
export const DELETE_PRODUCT_DETAILS = 'DELETE_PRODUCT_DETAILS';
export const DELETE_PRODUCT_FROM_LIST = 'DELETE_PRODUCT_FROM_LIST';
export const CLEAR_PRODUCTS = 'CLEAR_PRODUCTS';
export const FORM_START = 'FORM_START';
export const FORM_SUCCESS = 'FORM_SUCCESS';
export const FORM_FAIL = 'FORM_FAIL';
export const DATA_START = 'DATA_START';
export const DATA_SUCCESS = 'DATA_SUCCESS';
export const DATA_FAIL = 'DATA_FAIL';
export const SET_MESSAGE = 'SET_MESSAGE';
export const DELETE_MESSAGE = 'DELETE_MESSAGE';
export const SET_MODAL_OPEN_STATE = 'SET_MODAL_OPEN_STATE';
export const SET_MAX_QUANTITY_PER_PAGE = 'SET_MAX_QUANTITY_PER_PAGE';
