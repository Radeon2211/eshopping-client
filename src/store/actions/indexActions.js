export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions';

export { setModal, deleteMessage, changeProductsPerPage } from './uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions';
