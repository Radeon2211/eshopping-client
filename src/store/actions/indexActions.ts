export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeEmail,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions/authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions/tradeActions';

export { setMessage, setModal, changeProductsPerPage } from './uiActions/uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions/productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions/orderActions';
