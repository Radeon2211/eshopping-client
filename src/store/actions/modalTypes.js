export const SIGNUP = 'SIGNUP';
export const LOGIN = 'LOGIN';
export const ADD_PRODUCT = 'ADD_PRODUCT';
export const DELETE_PRODUCT = 'DELETE_PRODUCT';
export const CHANGE_NAME = 'CHANGE_NAME';
export const CHANGE_EMAIL = 'CHANGE_EMAIL';
export const CHANGE_USERNAME = 'CHANGE_USERNAME';
export const CHANGE_PHONE_NUMBER = 'CHANGE_PHONE_NUMBER';
export const CHANGE_ADDRESS = 'CHANGE_ADDRESS';
export const CHANGE_CONTACTS = 'CHANGE_CONTACTS';
