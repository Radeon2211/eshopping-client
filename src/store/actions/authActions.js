import axios from '../../axios';
import * as actionTypes from './actionTypes';
import * as uiActions from './uiActions';
import { getErrorMessage } from '../../shared/utility';
import { DEFAULT_PATH } from '../../shared/constants';

export const setProfile = (profile) => ({
  type: actionTypes.SET_PROFILE,
  profile,
});

export const logout = () => ({
  type: actionTypes.LOGOUT_USER,
});

export const setDeliveryAddress = (deliveryAddress) => ({
  type: actionTypes.SET_DELIVERY_ADDRESS,
  deliveryAddress,
});

export const setOtherUser = (user) => ({
  type: actionTypes.SET_OTHER_USER,
  otherUser: user,
});

export const registerUser = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    const country = creds.country.value;
    const phone = `+${creds.phonePrefix.value} ${creds.phoneNumber}`;
    const contacts = [];
    if (!creds.hideEmail) contacts.push('email');
    if (!creds.hidePhone) contacts.push('phone');
    const correctCreds = {
      ...creds,
      country,
      phone,
      contacts,
    };
    delete correctCreds.hideEmail;
    delete correctCreds.hidePhone;
    delete correctCreds.phoneNumber;
    delete correctCreds.phonePrefix;
    try {
      const { data } = await axios.post('/users', { data: correctCreds });
      dispatch(setProfile(data.user));
      dispatch(uiActions.setAndDeleteMessage('Your account has been created successfully!'));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const loginUser = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { data } = await axios.post('/users/login', { data: creds });
      dispatch(setProfile(data.user));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const fetchProfile = () => {
  return async (dispatch) => {
    try {
      const { data } = await axios.get('/users/me');
      dispatch(setProfile(data.user));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
    } catch (error) {
      dispatch(setProfile(null));
    }
  };
};

export const logoutUser = () => {
  return async (dispatch) => {
    try {
      await axios.post('/users/logout');
      dispatch(setProfile(null));
    } catch (error) {
      dispatch(uiActions.setAndDeleteMessage('Unable to logout. Something went wrong'));
    }
  };
};

export const updateUser = (creds, message) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { data } = await axios.patch('/users/me', creds);
      dispatch(setProfile(data.user));
      dispatch(uiActions.setAndDeleteMessage(message));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const deleteAccount = (creds, history) => {
  return async (dispatch, getState) => {
    dispatch(uiActions.formStart());
    try {
      const { username } = getState().auth.profile;
      await axios.delete('/users/me', { data: creds });
      dispatch(setProfile(null));
      dispatch(
        uiActions.setAndDeleteMessage(`Your account has been deleted. Goodbye ${username}!`),
      );
      dispatch(uiActions.formSuccess());
      history.replace(DEFAULT_PATH);
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const changeDeliveryAddress = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { onlyTheseOrders } = creds;
      // eslint-disable-next-line no-param-reassign
      delete creds.onlyTheseOrders;

      if (onlyTheseOrders) {
        dispatch(setDeliveryAddress(creds));
      } else {
        const { data } = await axios.patch('/users/me', creds);
        dispatch(setProfile(data.user));
        dispatch(uiActions.setAndDeleteMessage('Delivery address has been saved in your profile'));
      }
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const fetchOtherUser = (username) => {
  return async (dispatch) => {
    dispatch(uiActions.dataStart());
    try {
      const { data } = await axios.get(`/users/${username}`);
      dispatch(setOtherUser(data.profile));
      dispatch(uiActions.dataEnd());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(setOtherUser(null));
      dispatch(uiActions.dataEnd());
    }
  };
};
