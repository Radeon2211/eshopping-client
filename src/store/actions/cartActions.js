import axios from '../../axios';
import * as actionTypes from './actionTypes';
import * as uiActions from './uiActions';
import { getErrorMessage } from '../../shared/utility';
import { modalTypes } from '../../shared/constants';

let updateCartItemReqCounter = 0;

export const setCart = (cart) => ({
  type: actionTypes.SET_CART,
  cart,
});

export const fetchCart = () => {
  return async (dispatch) => {
    dispatch(uiActions.cartStart());
    try {
      const { data } = await axios.get('/cart');
      dispatch(uiActions.cartEnd());
      dispatch(setCart(data.cart));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(uiActions.cartEnd());
      dispatch(setCart(null));
    }
  };
};

export const addCartItem = (item) => {
  return async (dispatch) => {
    dispatch(uiActions.cartStart());
    try {
      const { data } = await axios.patch('/cart/add', item);
      dispatch(uiActions.cartEnd());
      dispatch(setCart(data.cart));
      dispatch(uiActions.setModal(true, modalTypes.CART_ITEM_ADDED));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(uiActions.cartEnd());
    }
  };
};

export const updateCartItem = (itemId, action, quantity) => {
  return async (dispatch) => {
    dispatch(uiActions.cartStart());
    try {
      updateCartItemReqCounter += 1;
      const quantityParam = quantity ? `&quantity=${quantity}` : '';
      const { data } = await axios.patch(`/cart/${itemId}/update?action=${action}${quantityParam}`);
      updateCartItemReqCounter -= 1;
      if (updateCartItemReqCounter <= 0) {
        dispatch(uiActions.cartEnd());
        dispatch(setCart(data.cart));
        dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
      }
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(uiActions.cartEnd());
    }
  };
};

export const clearCart = () => {
  return async (dispatch) => {
    dispatch(uiActions.cartStart());
    try {
      await axios.patch('/cart/clear');
      dispatch(uiActions.cartEnd());
      dispatch(setCart([]));
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(uiActions.cartEnd());
    }
  };
};

export const removeCartItem = (itemId) => {
  return async (dispatch) => {
    dispatch(uiActions.cartStart());
    try {
      const { data } = await axios.patch(`/cart/${itemId}/remove`);
      dispatch(uiActions.cartEnd());
      dispatch(setCart(data.cart));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(uiActions.cartEnd());
    }
  };
};
