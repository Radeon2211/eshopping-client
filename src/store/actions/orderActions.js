import queryString from 'query-string';
import axios from '../../axios';
import * as actionTypes from './actionTypes';
import * as uiActions from './uiActions';
import { orderTypes } from '../../shared/constants';
import { getErrorMessage, getParamsWithoutPollution } from '../../shared/utility/utility';

export const setPlacedOrders = (placedOrders, orderCount = undefined) => ({
  type: actionTypes.SET_PLACED_ORDERS,
  placedOrders,
  orderCount,
});

export const setSellHistory = (sellHistory, orderCount = undefined) => ({
  type: actionTypes.SET_SELL_HISTORY,
  sellHistory,
  orderCount,
});

export const setOrderDetails = (orderDetails) => ({
  type: actionTypes.SET_ORDER_DETAILS,
  orderDetails,
});

export const fetchOrders = (search, type) => {
  return async (dispatch) => {
    try {
      dispatch(uiActions.dataStart());
      const parsedQueryParams = getParamsWithoutPollution(search);
      const { p: pageNumber } = parsedQueryParams;

      if (!pageNumber || pageNumber < 1) parsedQueryParams.p = 1;

      const updatedQueryParams = queryString.stringify(parsedQueryParams);

      const {
        data: { orders, orderCount },
      } = await axios.get(`/orders?type=${type}&${updatedQueryParams}`);

      if (type === orderTypes.PLACED_ORDERS) {
        dispatch(setPlacedOrders(orders, orderCount));
      } else {
        dispatch(setSellHistory(orders, orderCount));
      }

      dispatch(uiActions.dataEnd());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(setPlacedOrders(null, undefined));
      dispatch(uiActions.dataEnd());
    }
  };
};

export const fetchOrderDetails = (orderId) => {
  return async (dispatch) => {
    try {
      dispatch(uiActions.dataStart());
      const { data } = await axios.get(`/orders/${orderId}`);
      dispatch(setOrderDetails(data.order));
      dispatch(uiActions.dataEnd());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(setOrderDetails(null));
      dispatch(uiActions.dataEnd());
    }
  };
};
