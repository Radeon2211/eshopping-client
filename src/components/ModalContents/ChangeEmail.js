import React, { useCallback } from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from '../../store/actions/indexActions';
import Form from '../UI/Form/Form';
import Input from '../UI/Input/Input';
import { inputKinds, userRules } from '../../shared/constants';
import PlainText from '../UI/PlainText';

const validationSchema = Yup.object({
  email: userRules.email,
});

const ChangeEmail = () => {
  const userProfile = useSelector((state) => state.auth.profile);

  const dispatch = useDispatch();
  const onChangeEmail = useCallback((creds) => dispatch(actions.changeEmail(creds)), [dispatch]);

  return (
    <Formik
      initialValues={{
        email: '',
        currentPassword: '',
      }}
      validationSchema={validationSchema}
      onSubmit={(data) => {
        onChangeEmail(data);
      }}
    >
      {({ dirty, errors, touched, values, isValid, setFieldTouched }) => (
        <Form
          btnText="change"
          headingText="Change your email"
          isValid={dirty && isValid && userProfile.email !== values.email}
          cancellable
        >
          <PlainText size="1" mgBottom="3" textAlign="justify">
            We will send to given email address an email with verification link. Link will be active
            for 10 minutes. To resend an email you will need to fill this form again.
          </PlainText>
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'email',
              name: 'email',
              id: 'email',
              placeholder: 'Your new email',
              autoComplete: 'email',
              autoFocus: true,
              onInput: setFieldTouched.bind(this, 'email', true, true),
            }}
            isValid={!errors.email && userProfile.email !== values.email}
            isTouched={touched.email}
            label="New email"
          />
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'password',
              name: 'currentPassword',
              id: 'currentPassword',
              placeholder: 'Your current password',
              autoComplete: 'off',
            }}
            label="Current password"
          />
        </Form>
      )}
    </Formik>
  );
};

export default ChangeEmail;
