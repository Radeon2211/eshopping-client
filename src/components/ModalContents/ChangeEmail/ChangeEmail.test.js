import { screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ChangeEmail from './ChangeEmail';
import {
  clickAtSubmitButton,
  defaultUserProfile,
  renderAppPart,
} from '../../../shared/testUtility/testUtility';
import * as actions from '../../../store/actions/indexActions';

const mockStore = configureMockStore([thunk]);

const oldEmail = 'old@domain.com';

const defaultStore = mockStore({
  auth: {
    profile: { ...defaultUserProfile, email: oldEmail },
  },
  ui: {
    isFormLoading: false,
    formError: '',
  },
});
defaultStore.dispatch = jest.fn();

const setUp = () => {
  return {
    ...renderAppPart(<ChangeEmail />, {
      store: defaultStore,
      withoutRouter: true,
    }),
    store: defaultStore,
  };
};

jest.mock('../../../store/actions/indexActions.js', () => ({
  ...jest.requireActual('../../../store/actions/indexActions.js'),
  changeEmail: (credentials) => credentials,
}));

describe('<ChangeEmail />', () => {
  describe('check form', () => {
    const defaultNewEmail = 'new@domain.com';
    const currentPassword = 'password';

    it('should have empty inputs and focus on email input by default', () => {
      setUp();
      const emailInput = screen.getByTestId('ChangeEmail-email');
      const currentPasswordInput = screen.getByTestId('ChangeEmail-current-password');
      expect(emailInput).toHaveFocus();
      expect(emailInput.value).toEqual('');
      expect(currentPasswordInput.value).toEqual('');
    });

    describe('should call', () => {
      it('should call changeEmail() with given credentials after input submit and button click', async () => {
        const { store, container } = setUp();

        const emailInput = screen.getByTestId('ChangeEmail-email');
        const currentPasswordInput = screen.getByTestId('ChangeEmail-current-password');

        const dataToPass = {
          email: defaultNewEmail,
          currentPassword,
        };

        await waitFor(() => {
          fireEvent.change(emailInput, { target: { value: defaultNewEmail } });
        });
        await waitFor(() => {
          fireEvent.change(currentPasswordInput, { target: { value: currentPassword } });
        });
        expect(emailInput.value).toEqual(defaultNewEmail);
        expect(currentPasswordInput.value).toEqual(currentPassword);

        await clickAtSubmitButton(container);
        expect(store.dispatch).toHaveBeenNthCalledWith(1, actions.changeEmail(dataToPass));

        await waitFor(() => {
          fireEvent.submit(emailInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(2, actions.changeEmail(dataToPass));

        expect(store.dispatch).toHaveBeenCalledTimes(2);
      });
    });

    describe('should NOT call', () => {
      it('should NOT call changeEmail() if both inputs are empty', async () => {
        const { store, container } = setUp();
        await clickAtSubmitButton(container);
        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call changeEmail() if current password input is empty (email is valid)', async () => {
        const { store, container } = setUp();

        const emailInput = screen.getByTestId('ChangeEmail-email');

        await waitFor(() => {
          fireEvent.change(emailInput, { target: { value: defaultNewEmail } });
        });
        expect(emailInput.value).toEqual(defaultNewEmail);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call changeEmail() if email input is empty (password is given)', async () => {
        const { store, container } = setUp();

        const currentPasswordInput = screen.getByTestId('ChangeEmail-current-password');

        await waitFor(() => {
          fireEvent.change(currentPasswordInput, { target: { value: currentPassword } });
        });
        expect(currentPasswordInput.value).toEqual(currentPassword);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call changeEmail() if email is invalid (password is given)', async () => {
        const { store, container } = setUp();

        const emailInput = screen.getByTestId('ChangeEmail-email');
        const currentPasswordInput = screen.getByTestId('ChangeEmail-current-password');

        const newEmail = 'invalidemail';

        await waitFor(() => {
          fireEvent.change(emailInput, { target: { value: newEmail } });
        });
        await waitFor(() => {
          fireEvent.change(currentPasswordInput, { target: { value: currentPassword } });
        });
        expect(emailInput.value).toEqual(newEmail);
        expect(currentPasswordInput.value).toEqual(currentPassword);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call changeEmail() if email is the same as current email (password is given)', async () => {
        const { store, container } = setUp();

        const emailInput = screen.getByTestId('ChangeEmail-email');
        const currentPasswordInput = screen.getByTestId('ChangeEmail-current-password');

        await waitFor(() => {
          fireEvent.change(emailInput, { target: { value: oldEmail } });
        });
        await waitFor(() => {
          fireEvent.change(currentPasswordInput, { target: { value: currentPassword } });
        });
        expect(emailInput.value).toEqual(oldEmail);
        expect(currentPasswordInput.value).toEqual(currentPassword);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });
    });
  });
});
