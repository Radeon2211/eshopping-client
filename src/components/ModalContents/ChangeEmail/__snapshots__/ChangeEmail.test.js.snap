// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangeEmail /> check how renders should render everything correctly 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  justify-self: center;
  max-width: 100%;
  text-align: left;
  width: 65rem;
}

.c11 {
  box-shadow: 0 0 3px rgba(0,0,0,0.03);
  border-radius: 1px;
  cursor: pointer;
  font-size: 1.4rem;
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  outline: none;
  padding: 0.8rem 1.2rem;
  position: relative;
  text-align: center;
  text-transform: uppercase;
  -webkit-transition: all 0.1s;
  transition: all 0.1s;
  background-color: transparent;
  border: 2px solid #1E8BC3;
  color: #1E8BC3;
}

.c11:hover {
  border-color: #22A7F0;
  color: #22A7F0;
}

.c12 {
  box-shadow: 0 0 3px rgba(0,0,0,0.03);
  border-radius: 1px;
  cursor: pointer;
  font-size: 1.4rem;
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  outline: none;
  padding: 0.8rem 1.2rem;
  position: relative;
  text-align: center;
  text-transform: uppercase;
  -webkit-transition: all 0.1s;
  transition: all 0.1s;
  background-color: #1E8BC3;
  border: 2px solid #1E8BC3;
  color: #fff;
  background-color: #67809F;
  border-color: #67809F;
  color: #FAFAFA;
  cursor: not-allowed;
}

.c12:hover {
  background-color: #22A7F0;
  border-color: #22A7F0;
}

.c12:hover {
  background-color: #67809F;
  border-color: #67809F;
  color: #FAFAFA;
}

.c2 {
  display: inline-block;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  margin-bottom: 2.4rem;
  margin-left: 0;
  margin-right: 0;
  overflow: visible;
  text-overflow: clip;
  white-space: normal;
  text-align: justify;
  font-size: 1.3rem;
}

.c10 {
  display: block;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  margin-bottom: 0;
  margin-left: 0;
  margin-right: 2.4rem;
  overflow: visible;
  text-overflow: clip;
  white-space: normal;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  width: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 > *:not(:last-child) {
  margin-right: 2.4rem;
}

.c1 {
  font-size: 2.4rem;
  font-family: Noto Sans,sans-serif;
  line-height: 1;
  margin-bottom: 2.4rem;
  margin-top: 0;
  text-align: left;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 2.4rem;
}

.c4 .select {
  font-size: 1.4rem;
}

.c4 .textarea {
  border: none;
  border-bottom: 3px solid rgba(0,0,0,0.1);
  outline: none;
  font-size: 1.4rem;
  font-family: inherit;
  padding: 0 0 0.8rem 0;
  -webkit-transition: border-bottom 0.1s;
  transition: border-bottom 0.1s;
  resize: none;
}

.c4 .textarea::-webkit-input-placeholder {
  color: #BFBFBF;
}

.c4 .textarea:focus {
  border-bottom: 3px solid rgba(0,0,0,0.35);
}

.c4.valid .textarea {
  border-bottom: 3px solid #009680;
}

.c4.invalid .textarea {
  border-bottom: 3px solid #F22613;
}

.c8 {
  border: none;
  border-bottom: 3px solid rgba(0,0,0,0.1);
  color: #000;
  font-size: 1.3rem;
  font-family: inherit;
  outline: none;
  padding: 0 0 0.8rem 0;
  -webkit-transition: border-bottom 0.1s;
  transition: border-bottom 0.1s;
  width: 100%;
}

.c8::-webkit-input-placeholder {
  color: #BFBFBF;
}

.c8:focus {
  border-bottom: 3px solid rgba(0,0,0,0.35);
}

.c3.valid .c7 {
  border-bottom: 3px solid #009680;
}

.c3.invalid .c7 {
  border-bottom: 3px solid #F22613;
}

.c6 {
  font-size: 1.5rem;
  padding-bottom: 0.8rem;
  -webkit-transition: color 0.1s;
  transition: color 0.1s;
}

.c3.valid .c5 {
  color: #009680;
}

.c3.invalid .c5 {
  color: #F22613;
}

<div
    class="c0"
  >
    <h3
      class="c1"
    >
      Change your email
    </h3>
    <form
      action="#"
      class="formik-form"
    >
      <span
        class="c2"
        color=""
        display="inline-block"
        overflow="visible"
        size="1"
        spacing="0px"
        transform=""
      >
        We will send to given email address an email with verification link. Link will be active for 10 minutes. To resend an email you will need to fill this form again.
      </span>
      <div
        class="c3 c4"
        type="email"
      >
        <label
          class="c5 c6"
          for="email"
        >
          New email
        </label>
        <input
          autocomplete="email"
          class="c7 c8"
          data-testid="ChangeEmail-email"
          id="email"
          name="email"
          placeholder="Your new email"
          type="email"
          value=""
        />
      </div>
      <div
        class="c3 c4"
        type="password"
      >
        <label
          class="c5 c6"
          for="currentPassword"
        >
          Current password
        </label>
        <input
          autocomplete="off"
          class="c7 c8"
          data-testid="ChangeEmail-current-password"
          id="currentPassword"
          name="currentPassword"
          placeholder="Your current password"
          type="password"
          value=""
        />
      </div>
      <div
        class="c9"
        direction="row"
        spacing="3"
        width="100%"
        wrap="nowrap"
      >
        <span
          class="c10"
          color=""
          display="block"
          overflow="visible"
          spacing="0px"
          transform=""
        >
          <button
            class="c11"
            color="blue"
            type="button"
          >
            cancel
          </button>
        </span>
        <button
          class="c12"
          color="blue"
          data-testid="Form-submit-btn"
          disabled=""
          type="submit"
        >
          change
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
